### ISSUE TITLE
Support JDK9 JEP 238: Multi-Release JAR Files

### ISSUE THREAD

**Issue by ktoso** [sbt/sbt#3036]
We're looking to adopt JDK9 rather early in order to provide Akka users additional performance by using the new fences and var handle APIs etc.

In order to manage that as a clean transition and not complete fork of Akka we'll need Multi-Release JARs that are coming in in JDK9: http://openjdk.java.net/jeps/238

It is rather simple AFAICS and boils down to putting classes under META-INF if they are meant to be "only jdk9" etc:

```
jar root
  - A.class
  - B.class
  - C.class
  - D.class
  - META-INF
     - versions
        - 9
           - A.class
           - B.class
```

Is this something you have looked at or not yet, if not yet and not planning to I'm willing to help out (work on it, may need some hints).
Do you think this is something that would fit well to provide in sbt itself and not as a plugin? 

I thought it would be good to provide this in sbt itself since it's a JDK thing - WDYT?

**Comment by dwijnand** [comment]
Being a JDK thing we'd definitely be interested in making sure it's supported, or at the very least sbt isn't an impediment to it.

Whether or not to build it into sbt or have it as a plugin depends on, from your assessment, whether you see Multi-Release JARs as (a) general purpose or (b) a niche concern.

If it's niche we'd still be interested in the necessary changes to provide hooks for such an external plugin.

**Comment by ktoso** [comment]
Ok that makes sense. Could you point me to where the thing is implemented
for multiple scala versions, because that's built in nowadays right?

The thing that i can say 2.12 and those sources only are used in 2.12.

**Comment by ktoso** [comment]
Hmm... I guess actually it may relate more to the packaging plugin and not sbt itself hm WDYT @eed3si9n ?

**Comment by dwijnand** [comment]
> The thing that i can say 2.12 and those sources only are used in 2.12. 

https://github.com/sbt/sbt/pull/1799

**Comment by dwijnand** [comment]
.. and there @eed3si9n points to `crossPaths` and `crossTarget`. That should start pointing you to the right places.

**Comment by ktoso** [comment]
Ok thanks, will poke around it on airports then 👍

**Comment by jvican** [comment]
I think this would be great @ktoso. Glad to hear you're working on it.

**Comment by ktoso** [comment]
Status update: not currently working on it, will report back if on it.

**Comment by ktoso** [comment]
Status update, we'll definitely need it very soon so will look into it I guess

**Comment by jroper** [comment]
Can this feature be utilised for cross building Scala as well?

**Comment by ktoso** [comment]
I don't think so, it's hardcoded to know about JDK versions, nothing else. It can handle `/META-INF/versions/9`, `/META-INF/versions/10` etc, but there's no way to smuggle in scala versions there AFAICS

**Comment by jroper** [comment]
That's a pity.

**Comment by ktoso** [comment]
It's something we could perhaps bring up during our discussions with them - @retronym FYI.
I guess they did the minimal thing they needed for 9, rather than doing a full API around it since time pressure to ship 9.

**Comment by retronym** [comment]
[Blog](http://in.relation.to/2017/02/13/building-multi-release-jars-with-maven/) about how to configure this with Maven.

**Comment by ktoso** [comment]
I have done so in a strawman proposal for Akka https://github.com/ktoso/akka-streams-jdk9-strawman-mvn :)

**Comment by ktoso** [comment]
I'm working on it FYI @retronym @dwijnand @eed3si9n 
Starting as a plugin: https://github.com/sbt/sbt-multi-release-jar

I hope to have it complete by the end of this week. Fingers crossed there won't be any roadblocks to uncover.
